### PR TITLE
Add support for elements with negative lengths

### LINKED ISSUE #448: [Feature Request] Handling Backtracking and Elements with negative lengths
It's nice to have the capability of backtracking in Cheetah. 

For most elements with linear transfer maps, this can be done simply by setting the lengths to the negative value (This is currently allowed in Cheetah). See the below snippet for an example. 

We could add this in the documentation, and add properties in elements specifying whether they support backtracking. 
We can also add a convenient method to revert a lattice.

```python
import cheetah
import torch
# Define a simple FODO lattice
fodo_lattice = cheetah.Segment(
    [
        cheetah.Quadrupole(length=torch.tensor(0.5), k1=torch.tensor(1.0), dtype=torch.float64, name="q1"),
        cheetah.Drift(length=torch.tensor(1.0), dtype=torch.float64, name="d1"),
        cheetah.Quadrupole(length=torch.tensor(0.5), k1=torch.tensor(-1.0), dtype=torch.float64, name="q2"),
        cheetah.Drift(length=torch.tensor(1.0), dtype=torch.float64, name="d2"),
    ]
)

# Define incoming beam parameters
incoming = cheetah.ParticleBeam.from_parameters(
    num_particles=1000,
    energy=torch.tensor(1e9),
    mu_x=torch.tensor(0.0),
    sigma_x=torch.tensor(1e-5),
    mu_px=torch.tensor(1e-3),
    dtype=torch.float64,
)
# Track the incoming beam through the lattice
outbeam = fodo_lattice(incoming)

# Create backtracking segment
reversed_elements = []
for element in fodo_lattice.elements[::-1]:  # revert the elements order
    reversed_element = element.clone()
    reversed_element.length = -element.length  # backtrack through elements 
    reversed_elements.append(reversed_element)
reversed_lattice = cheetah.Segment(reversed_elements)
# Track the outgoing beam back through the reversed lattice
backtracked_incoming = reversed_lattice(outbeam)
# Check if the backtracked beam matches the original incoming beam
assert not torch.allclose(incoming.particles, outbeam.particles)
assert torch.allclose(backtracked_incoming.particles, incoming.particles)
```

In addition, some simulation codes like elegant utilizes the negative length drift to define overlapping accelerator components. E.g. a marker in the quadrupole, 

```python
# Example lattice containing a negative length drift
fodo_lattice_neg_drift = cheetah.Segment(
    [
        cheetah.Quadrupole(length=torch.tensor(0.5), k1=torch.tensor(1.0), dtype=torch.float64, name="q1"),
        cheetah.Drift(length=torch.tensor(-0.25), dtype=torch.float64, name="d_q1_m1"),
        cheetah.Marker(name="marker1"),
        cheetah.Drift(length=torch.tensor(0.25), dtype=torch.float64, name="d_m1_q1"), # Drift back to the end of q1
        cheetah.Drift(length=torch.tensor(1.0), dtype=torch.float64, name="d1"),
        cheetah.Quadrupole(length=torch.tensor(0.5), k1=torch.tensor(-1.0), dtype=torch.float64, name="q2"),
        cheetah.Drift(length=torch.tensor(1.0), dtype=torch.float64, name="d2"),  # Negative length drift
    ]
)
```

Allowing negative lengths element also introduces some problem in other convenience methods, e.g. plot overview:

```python
# Define some incoming beam parameters
incoming = cheetah.ParticleBeam.from_parameters(
    num_particles=1000,
    energy=torch.tensor(1e9),
    mu_x=torch.tensor(0.0),
    sigma_x=torch.tensor(1e-5),
    mu_px=torch.tensor(0.0),
    sigma_px = torch.tensor(1e-5),
    mu_y=torch.tensor(0.0),
    sigma_y=torch.tensor(1e-5),
    mu_py=torch.tensor(0.0),
    sigma_py=torch.tensor(1e-5),
    dtype=torch.float64,
)
# Example lattice containing a negative length drift
fodo_lattice_neg_drift = cheetah.Segment(
    [
        cheetah.Quadrupole(length=torch.tensor(0.5), k1=torch.tensor(1.0), dtype=torch.float64, name="q1"),
        cheetah.Drift(length=torch.tensor(-0.25), dtype=torch.float64, name="d_q1_m1"),
        cheetah.Marker(name="marker1"),
        cheetah.Drift(length=torch.tensor(0.25), dtype=torch.float64, name="d_m1_q1"), # Drift back to the end of q1
        cheetah.Drift(length=torch.tensor(1.0), dtype=torch.float64, name="d1"),
        cheetah.Quadrupole(length=torch.tensor(0.5), k1=torch.tensor(-1.0), dtype=torch.float64, name="q2"),
        cheetah.Drift(length=torch.tensor(1.0), dtype=torch.float64, name="d2"),  # Negative length drift
    ]
)
# The same lattice without the inserted marker and negative length drift
fodo_lattice = cheetah.Segment(
    [
        cheetah.Quadrupole(length=torch.tensor(0.5), k1=torch.tensor(1.0), dtype=torch.float64, name="q1"),
        cheetah.Drift(length=torch.tensor(1.0), dtype=torch.float64, name="d1"),
        cheetah.Quadrupole(length=torch.tensor(0.5), k1=torch.tensor(-1.0), dtype=torch.float64, name="q2"),
        cheetah.Drift(length=torch.tensor(1.0), dtype=torch.float64, name="d2"),  # Negative length drift
    ]
)
fodo_lattice_neg_drift.plot_overview(incoming)
fodo_lattice.plot_overview(incoming)
```

With negative drift, notice that the s-positions seem to be messed up:

Image

What it should look like 

Image

### PR DESCRIPTION
**Pull request by Hespe**
## Description

In some situations, it is reasonable to define negative length elements, especially `Drift`. This PR ensures we are compatible and tracking does not fail in unexpected ways.

## Motivation and Context

- [x] I have raised an issue to propose this change (required for new features and bug fixes)
  - Closes #448

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (update in the documentation)

## Checklist

- [x] I have updated the changelog accordingly (**required**).
- [ ] My change requires a change to the documentation.
- [x] I have updated the tests accordingly (_required for a bug fix or a new feature_).
- [ ] I have updated the documentation accordingly.
  
  
- [x] I have reformatted the code and checked that formatting passes (**required**).
- [x] I have have fixed all issues found by `flake8` (**required**).
- [x] I have ensured that all `pytest` tests pass (**required**).
- [ ] I have run `pytest` on a machine with a CUDA GPU and made sure all tests pass (**required**).
- [x] I have checked that the documentation builds (**required**).

Note: We are using a maximum length of 88 characters per line.

### CODE REVIEW

**Comment by jank324** [comment]
> This test failure is strange. I cannot reliably reproduce it locally but it happens once in a while also on my computer. However, `SpaceChargeKick` does not even implement `split` other than returning `[self]`, so there should be absolutely no difference after splitting.

Did you try running `pytest` with the same random seed? Or can you not even reproduce it then? I would suspect that the tracking through a `SpaceChargeKick` is somehow stochastic. Still weird, because we didn't really touch that in a long time.

**Comment by Hespe** [comment]
> Did you try running `pytest` with the same random seed? Or can you not even reproduce it then? I would suspect that the tracking through a `SpaceChargeKick` is somehow stochastic. Still weird, because we didn't really touch that in a long time.

Even with fixed random seed that failed on CI the test sometimes succeeds locally. But it fails most of the time, at least.

**Comment by jank324** [comment]
> > Did you try running `pytest` with the same random seed? Or can you not even reproduce it then? I would suspect that the tracking through a `SpaceChargeKick` is somehow stochastic. Still weird, because we didn't really touch that in a long time.
> 
> Even with fixed random seed that failed on CI the test sometimes succeeds locally. But it fails most of the time, at least.

Odd ... I guess that would imply we are not actually seed everything.

**Comment by Hespe** [comment]
> Odd ... I guess that would imply we are not actually seeding everything.

The answer is likely in https://docs.pytorch.org/docs/stable/notes/randomness.html#avoiding-nondeterministic-algorithms. I have observed that `SpaceChargeKick` returns slightly different beams when tracking the same incoming beam even though it should have no internal state and is using no randomness.

Since setting `torch.use_deterministic_algorithms(True)` is slowing down PyTorch, I do not think that we should apply it generaly but leave that choice to the user (if they care at all). But to save us some trouble when running the automated tests, I have configured pytest to always set this flag when seeding the PRNGs such that we should hopefully not observe any random failures of `SpaceChargeKick` from now on.

**Comment by jank324** [comment]
> @jank324 I was not sure if we should explicitly add the removal of the broken custom `split` implementation in `HorizontalCorrector` and `VerticalCorrector` to the changelog. So far I did not.

I would add it as a bug fix.

**Review by jank324 — APPROVED** [review]
